src/lib.rs

```diff
@@ -680,13 +705,43 @@ mod test {
         let result = select_coin_srd(&inputs, options);
         assert!(matches!(result, Err(SelectionError::InsufficientFunds)));
     }
-    fn test_nonexact_match_knapsack() {
-        let mut inputs = setup_knapsack_output_groups();
-        let mut options = setup_options(1500); //Set target value to match the available utxos, target fee rate, base weight
-        let mut result = select_coin_knapsack(&inputs, options);
-        assert!(result.is_ok());
-        let mut selection_output = result.unwrap();
-        assert!(!selection_output.selected_inputs.is_empty());
+    fn test_core_knapsack_vectors() {
+        for i in 0..RUN_TESTS {
+            // Test if Knapsack retruns an Error
+            let mut inputs: Vec<OutputGroup> = Vec::new();
+            let mut options = setup_options(1000);
+            let mut result = select_coin_knapsack(&inputs, options);
+            assert!(matches!(result, Err(SelectionError::NoSolutionFound)));
+
+            // Adding 2 CENT and 1 CENT to the wallet and testing if knapsack can select the two inputs for a 3 CENT Output
+            // Adding 0.001 CENT to the two inputs to account for fees.
+            inputs = setup_output_groups(
+                vec![(2.001 * CENT).round() as u64, (1.001 * CENT).round() as u64],
+                vec![30, 100],
+            );
+            options = setup_options((3.0 * CENT).round() as u64);
+            if let Ok(result) = select_coin_knapsack(&inputs, options) {
+                assert_eq!(result.selected_inputs.len(), 2);
+            }
+            // Clearing the input vector
+            inputs.clear();
+            // Adding 2 CENT and 1 CENT to the wallet and testing if knapsack can select the two inputs for a 3 CENT Output
+            // Adding 0.001 CENT to the two inputs to account for fees.
+            inputs = setup_output_groups(
+                vec![
+                    (2.001 * CENT).round() as u64,
+                    (1.001 * CENT).round() as u64,
+                    (5.001 * CENT).round() as u64,
+                    (10.001 * CENT).round() as u64,
+                    (20.001 * CENT).round() as u64,
+                ],
+                vec![30, 100, 100, 50, 40],
+            );
+            options = setup_options((37.0 * CENT).round() as u64);
+            if let Ok(result) = select_coin_knapsack(&inputs, options) {
+                assert_eq!(result.selected_inputs.len(), 5);
+            }
```

Comment:
In some iterations, it selects 4 inputs, while in some it selects all the 5 inputs. Hence the tests fail. Ideally we expect it to select, 20, 10, 5 & 2.